Knights fight v.1

Filllick opened the pull request.

vasylhnatiuk commented on app/main.py:
```diff
@@ -86,129 +90,24 @@
+        ("Arthur", "Red Knight")
+    ]
+
+    for knight1 in prepared_heroes:
```
you can avoid two loops here, create dict in prepare_knights where kinght name is a key, and knight instance is value
and after this you can get instance by name

vasylhnatiuk commented on app/Challengers/Knight.py:
```diff
@@ -0,0 +1,21 @@
+
+    @hp.setter
+    def hp(self, value: int) -> None:
+        if value < 0:
```
you can do it in one line

Filllick commented:
Changed list_knights function to dict-based knights_to_dict and prepare_knights

vsmutok commented on app/battle/preparation.py:
```diff
@@ -0,0 +1,38 @@
+
+    @staticmethod
+    def prepare_knights(
+            knights_to_dict: dict,
```
you already have a method with that name, using the same names is bad practice

vsmutok reviewed (changes requested):
folder names should also be written in lower case

Filllick commented:
renamed prepare_knights parameter from knights_to_dict to dict_of_knights

vsmutok commented on app/battle/preparation.py:
```diff
@@ -0,0 +1,38 @@
+        ) for data in knight_data.values()
+        }
+
+    @staticmethod
+    def prepare_knights(
+            dict_of_knights: dict,
+            knights_data: dict
+    ) -> dict:
+
+        for fighter in knights_data.values():
+            fighter["armour"] = sum(
+                part["protection"] for part in fighter["armour"]
+            )
+            fighter["power"] += fighter["weapon"]["power"]
+
+            if fighter["potion"]:
+                potion_effect = fighter["potion"]["effect"]
+                fighter["power"] += potion_effect.get("power", 0)
+                fighter["hp"] += potion_effect.get("hp", 0)
+                fighter["armour"] += potion_effect.get("protection", 0)
+
+        for knight in dict_of_knights.values():
+            for data in knights_data.values():
+                if knight.name == data["name"]:
+                    knight.armour = data["armour"]
+                    knight.power = data["power"]
+                    knight.hp = data["hp"]
+
+        return dict_of_knights
```
I think you can move this to class Hero

Filllick commented:
Мoved prepare_knights to Hero class.
But following the task, I created module Knight just to create fighters stats and check if fighters hp >=0, then created preparation module, with it`s methods fighters prepare (they sign up for the tournament with knights_to_dict method and equip for the fight with prepare_knights method) and the duel module that simulates the fight of two fighters.

ArsenPidhoretskyi commented on app/challengers/Knight.py:
```diff
@@ -0,0 +1,45 @@
+
+    @property
+    def hp(self) -> int:
+        return self._hp
```
You do not have this attribute. Add it to init

vasylhnatiuk replied:
do it

ArsenPidhoretskyi commented on app/challengers/Knight.py:
```diff
@@ -0,0 +1,45 @@
+
+            if fighter["potion"]:
+                potion_effect = fighter["potion"]["effect"]
+                fighter["power"] += potion_effect.get("power", 0)
+                fighter["hp"] += potion_effect.get("hp", 0)
+                fighter["armour"] += potion_effect.get("protection", 0)
```
Use forloop

ArsenPidhoretskyi commented on app/battle/preparation.py:
```diff
@@ -0,0 +1,11 @@
+from app.challengers.Knight import Hero
+
+
+class PreparedKnights:
```
This class is so strange and redundant

Filllick commented:
Moved knights_to_dict method do a Hero class, added forloop to the potion applying logic, but have no idea about this:
    @property
    def hp(self) -> int:
        return self._hp
hp is already in Hero init

Filllick commented:
added self._hp = hp to Hero class
